# Maize-sc-ATAC-seq
Identify novel genes/alleles/QTLs (high KRN with no or less off types) for improving KRN and yield via single cell ATAC-seq data analysis
## reformat.py
Python script to reformt and split the all tissue single cell ATAC-seq data from GEO into tissue specific and ArchR readable format 
Run this script first.
## sc-ATAC-seq_analysis.R
R script for single cell ATAC-seq data processing and figure plotting 
