# Maize-sc-ATAC-seq
Identify novel genes/alleles/QTLs (high KRN with no or less off types) for improving KRN and yield via single cell ATAC-seq data analysis
## reformat.py
Python script to reformt and split the all tissue single cell ATAC-seq data from GEO into tissue specific and ArchR readable format.

Run this script first.
 
## all_tissue_sc-atac-seq.R
R script for all tissue data from (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE155178) processing and figure plotting.
## ear_tissue_sc-atac-seq.R
R script for ear tissue data from (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE155178) processing and figure plotting.
